### PR TITLE
Feat gitlab support

### PR DESCRIPTION
**Pull request by jsdidierlaurent**

### CODE REVIEW

File: monitorables/gitlab/api/models/issuesparams.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+package models
+
+import "github.com/monitoror/monitoror/internal/pkg/monitorable/params"
+
+type (
+	IssuesParams struct {
+		params.Default
+
+		ProjectID *int `json:"projectId" query:"projectId"`
```
Comment (Alex-D) [inline monitorables/gitlab/api/models/issuesparams.go:11]: Not required?

File: monitorables/gitlab/api/usecase/gitlab.go [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+	"github.com/jsdidierlaurent/echo-middleware/cache"
+)
+
+type (
+	gitlabUsecase struct {
+		repository api.Repository
+		// Used to generate store key by repository
+		repositoryUID string
+
+		// store is used to store persistent data (project,
```
Comment (Alex-D) [inline monitorables/gitlab/api/usecase/gitlab.go:28]: > project,

??? what's next? :D

File: monitorables/gitlab/api/usecase/gitlab.go [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+	case "canceled":
+		return coreModels.CanceledStatus
+	case "skipped":
+		return coreModels.CanceledStatus
+	case "created":
+		return coreModels.QueuedStatus
+	case "manual":
+		return coreModels.ActionRequiredStatus
+	default:
+		return coreModels.UnknownStatus
```
Comment (Alex-D) [inline monitorables/gitlab/api/usecase/gitlab.go:287]: Maybe group keys that return the same value?
Is there a way in golang to get multiple cases with the same return?

File: docs/documentation/index.html [hunk 4]
```diff
@@ -1217,6 +1225,323 @@ <h5 class="m-documentation--configuration-side-title">UI configuration</h5>
...
+        <dd>
+          GitLab project ID (from project setting page)
+        </dd>
+
+        <dt><code>state</code> <code class="type">string</code></dt>
+        <dd>
+          GitLab issue state
+        </dd>
+
+        <dt><code>labels</code> <code class="type">[]string</code></dt>
```
Comment (Alex-D) [inline docs/documentation/index.html:1301]: string[]

File: docs/documentation/index.html [hunk 5]
```diff
@@ -1217,6 +1225,323 @@ <h5 class="m-documentation--configuration-side-title">UI configuration</h5>
...
+        </dd>
+
+        <dt><code>scope</code> <code class="type">string</code></dt>
+        <dd>
+          GitLab issue scope
+        </dd>
+
+        <dt><code>search</code> <code class="type">string</code></dt>
+        <dd>
+          Keyword to search in the title or description
```
Comment (Alex-D) [inline docs/documentation/index.html:1318]: Only one keyword is allowed?

File: docs/documentation/index.html [hunk 6]
```diff
@@ -1217,6 +1225,323 @@ <h5 class="m-documentation--configuration-side-title">UI configuration</h5>
...
+        </dd>
+
+        <dt><code>authorId</code> <code class="type">number</code></dt>
+        <dd>
+          GitLab user ID (from user setting page)
```
Comment (Alex-D) [inline docs/documentation/index.html:1323]: setting**s**

File: docs/documentation/index.html [hunk 7]
```diff
@@ -1217,6 +1225,323 @@ <h5 class="m-documentation--configuration-side-title">UI configuration</h5>
...
+        </dd>
+
+        <dt><code>assigneeId</code> <code class="type">number</code></dt>
+        <dd>
+          GitLab user ID (from user setting page)
```
Comment (Alex-D) [inline docs/documentation/index.html:1328]: setting**s**

File: docs/documentation/index.html [hunk 8]
```diff
@@ -1217,6 +1225,323 @@ <h5 class="m-documentation--configuration-side-title">UI configuration</h5>
...
+      <p>
+        Show a counter of issues.
+      </p>
+
+      <h5 class="m-documentation--configuration-side-title">UI configuration</h5>
+
+      <dl>
+        <dt><code>projectId</code> <code class="type">number</code></dt>
+        <dd>
+          GitLab project ID (from project setting page)
```
Comment (Alex-D) [inline docs/documentation/index.html:1293]: setting**s**

File: docs/documentation/index.html [hunk 9]
```diff
@@ -1217,6 +1225,323 @@ <h5 class="m-documentation--configuration-side-title">UI configuration</h5>
...
+  "params": {
+    "projectId": 42,
+    "id": 10
+  }
+}
+        </code></pre>
+
+        <div class="m-documentation--demo">
+          <div class="m-documentation--demo-tile">
+            <div class="m-documentation--demo-label">PR#10 @ monitoror</div>
```
Comment (Alex-D) [inline docs/documentation/index.html:1471]: MR

File: monitorables/gitlab/api/models/issuesparams.go (hunk 1 above)
Comment (jsdidierlaurent) [reply monitorables/gitlab/api/models/issuesparams.go:11]: According gitlab api documentation you can query issues with and without project ID:
- https://docs.gitlab.com/ee/api/issues.html#list-issues
- https://docs.gitlab.com/ee/api/issues.html#list-project-issues

File: docs/documentation/index.html (hunk 5 above)
Comment (jsdidierlaurent) [reply docs/documentation/index.html:1318]: Don't know, gitlab api say "string" :
https://docs.gitlab.com/ee/api/issues.html#list-project-issues

File: monitorables/github/api/repository/api.go [hunk 10]
```diff
@@ -178,10 +184,15 @@ func (gr *githubRepository) GetCommit(owner, repository, sha string) (*models.Co
...
 		SHA: sha,
-		Author: &coreModels.Author{
-			Name:      commit.Author.GetLogin(),
-			AvatarURL: gravatar.GetGravatarURL(commit.Author.GetEmail()),
-		},
+	}
+
+	if commit.GetAuthor() != nil {
+		result.Author.Name = commit.GetAuthor().GetName()
+		result.Author.AvatarURL = gravatar.GetGravatarURL(commit.GetAuthor().GetEmail())
```
[the remark this answers is not in the document]
Comment (jsdidierlaurent) [reply monitorables/github/api/repository/api.go:191]: It's not the same Struct.

```
func (p *PullRequest) GetUser() *User { ... } => https://github.com/google/go-github/blob/master/github/users.go#L20

func (c *Commit) GetAuthor() *CommitAuthor { ... } => https://github.com/google/go-github/blob/e4b511744f5bbb020d00e172a284468137d4e5fc/github/git_commits.go#L58
```
User have Avatar URL but CommitAuthor only have Email

File: monitorables/gitlab/api/usecase/gitlab.go (hunk 3 above)
Comment (jsdidierlaurent) [reply monitorables/gitlab/api/usecase/gitlab.go:286]: I can, but i don't like it. It's not git diff friendly.

Comment (jsdidierlaurent) [reply monitorables/gitlab/api/usecase/gitlab.go:286]: look's like : 

```
func parseStatus(status string) coreModels.TileStatus {
	// See: https://docs.gitlab.com/ee/api/pipelines.html#list-project-pipelines
	switch status {
	case "running":
		return coreModels.RunningStatus
	case "pending", "created":
		return coreModels.QueuedStatus
	case "success":
		return coreModels.SuccessStatus
	case "failed":
		return coreModels.FailedStatus
	case "canceled", "skipped":
		return coreModels.CanceledStatus
	case "manual":
		return coreModels.ActionRequiredStatus
	default:
		return coreModels.UnknownStatus
	}
}```

File: monitorables/github/api/repository/api.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (Alex-D) [reply monitorables/github/api/repository/api.go:191]: OK :)

File: monitorables/gitlab/api/models/issuesparams.go (hunk 1 above)
Comment (Alex-D) [reply monitorables/gitlab/api/models/issuesparams.go:11]: OK :)

File: monitorables/gitlab/api/usecase/gitlab.go (hunk 3 above)
Comment (Alex-D) [reply monitorables/gitlab/api/usecase/gitlab.go:286]: +1 for the git diff
what about just group them?

File: docs/documentation/index.html (hunk 5 above)
Comment (Alex-D) [reply docs/documentation/index.html:1318]: In the example:

> issue+title+or+description

so, maybe we can support a space-separated list of keywords? :)

Comment (jsdidierlaurent) [reply docs/documentation/index.html:1318]: it's already like that, if you use +, they are escaped and it does not work

File: monitorables/gitlab/api/usecase/gitlab.go (hunk 3 above)
Comment (jsdidierlaurent) [reply monitorables/gitlab/api/usecase/gitlab.go:286]: I use the ordrer of this page : https://docs.gitlab.com/ee/api/pipelines.html#list-project-pipelines

File: docs/documentation/index.html (hunk 5 above)
Comment (Alex-D) [reply docs/documentation/index.html:1318]: So change the docs to reflect that:

"Space-separated keywords to search in the title or description"

File: monitorables/gitlab/api/usecase/gitlab.go (hunk 3 above)
Comment (Alex-D) [reply monitorables/gitlab/api/usecase/gitlab.go:286]: OK

**Pull request merged**